libnd4j/include/loops/cpu/type_conversions.cpp

```diff
@@ -88,36 +88,96 @@ namespace nd4j {
         }
     }
 
+    /**
+ * This is cpu version, so leave it here as inline, to avoid templates instantiation
+ *
+ * @tparam S source type
+ * @tparam T target type
+ * @tparam T intermediate type
+ * @param dx
+ * @param N
+ * @param dz
+ */
+    template<typename S, typename T>
+    void TypeCast::convertDirectGeneric(Nd4jPointer * extras, void *dx, Nd4jLong N, void *dz) {
+        auto x = reinterpret_cast<S *>(dx);
+        auto z = reinterpret_cast<T *>(dz);
+
+        if (N < nd4j::Environment::getInstance()->elementwiseThreshold()) {
+#pragma omp simd
+            for (int i = 0; i < N; i++) {
+                z[i] = static_cast<T>(x[i]);
+            }
+        } else {
+
+#pragma omp parallel for
+            for (int i = 0; i < N; i++) {
+                z[i] = static_cast<T>(x[i]);
+            }
+        }
+    };
+
     /**
      * This is cpu version, so leave it here as inline, to avoid templates instantiation
      *
-     * @tparam S
-     * @tparam T
+     * @tparam S source type
+     * @tparam T target type
+     * @tparam T intermediate type
      * @param dx
      * @param N
      * @param dz
      */
-    template<typename S, typename T>
+    template<typename S, typename T, typename I>
     void TypeCast::convertGeneric(Nd4jPointer * extras, void *dx, Nd4jLong N, void *dz) {
         auto x = reinterpret_cast<S *>(dx);
         auto z = reinterpret_cast<T *>(dz);
 
         if (N < nd4j::Environment::getInstance()->elementwiseThreshold()) {
 #pragma omp simd
             for (int i = 0; i < N; i++) {
-                // FIXME: get rid of through-float though
-                z[i] = static_cast<T>(static_cast<float>(x[i]));
+                // FIXME: get rid of through-I though
+                z[i] = static_cast<T>(static_cast<I>(x[i]));
             }
         } else {
 
 #pragma omp parallel for
             for (int i = 0; i < N; i++) {
-                // FIXME: get rid of through-float though
-                z[i] = static_cast<T>(static_cast<float>(x[i]));
+                // FIXME: get rid of through-I though
```

Comment:
All this stuff is NOT need. 
Reason of that FIXME message isn't related to precision.

Reply:
In ideal world we should just add proper conversions between values, plus something for overflow handling. But definitely not a third template argument.

Reply:
So, each of defined additional dtype should have conversion operator defined for all other additional dtypes. That's proper solution we want long-term.

Reply:
Ok, I think I understand.

So if I understand you correctly 
`z[i] = static_cast<T>(static_cast<float>(x[i]));` does not result in a downcast even if `x[i]` is a double and 'z[i]' is a long long int.

Also, can you tell me whether the compiler takes care of optimizing 
`static_cast<float>(static_cast<float>(x[i]))` ? Or maybe we simple do not care.


If we were to get rid of the cast-through dtype, that would require implementing more conversion operators in:
```
float8.cpp
float16.cpp
float16.cpp
float16.cpp

Reply:
re: methods. Yes, exactly. those classes.

re: cast-through. problem is precision, not a capacity. i.e. you cast from **int** to **long** via float. In this case due to how floating point works, you'll get inaccurate conversions above some threshold.